File: bot/src/command/HelpCommand.ts
import Command from "./Command";
import {Context} from "../Context";
import {VkMessage} from "../service/VkMessagesService";
import {version} from "../main";

export default class HelpCommand extends Command {
    constructor(context: Context) {
        super(context);
    }

    getCommandShortUsage(): string {
        return '/sweet help';
    }

    canYouHandleThisCommand(command: string, message: VkMessage): boolean {
        return command === 'help' || command === '';
    }

    async handle(command: string, rawArguments: string, message: VkMessage): Promise<void> {
        const { vkMessagesService, botService, userPermissionsService } = this.context;
        let response = '';
        response += `Sweetie AI v${version}\n\n`;
        response += `Вот что можно сделать:\n`
        botService.getCommandHandlers()
            .filter(command => !command.chatAdminOnly(message.peerId) && !command.appCeoOnly())
            .forEach(command => {
                response += `${command.getCommandShortUsage()}\n`
            });
        if (await userPermissionsService.isUserPrivileged(message.peerId, message.fromId)) {
            response += `\nАдминские команды:\n`
            botService.getCommandHandlers()
                .filter(command => command.chatAdminOnly(message.peerId) && !command.appCeoOnly())
                .forEach(command => {
                    response += `${command.getCommandShortUsage()}\n`
                });
            response += `/sweet enable`;
        }
        await vkMessagesService.send(message.peerId, response);
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/command/HelpCommand.kt
package space.davids_digital.sweetie.command

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.integration.vk.VkMessagesService
import space.davids_digital.sweetie.model.VkMessageModel

@Component
class HelpCommand(
    @Qualifier("appVersion")
    private val appVersion: String,
    private val commandRegistry: CommandRegistry,
    private val vkMessagesService: VkMessagesService
): Command {
    override fun getNames(): Array<String> {
        return arrayOf("help", "")
    }

    override fun getUsage(): String {
        return "help"
    }

    override fun requiresChatAdmin(): Boolean {
        return false
    }

    override fun requiresAppCeo(): Boolean {
        return false
    }

    override suspend fun handle(commandName: String, rawArguments: String, message: VkMessageModel) {
        val builder = StringBuilder()
        builder.append("Sweetie AI v$appVersion\n")
        builder.append("Вот что можно сделать:\n")
        val userCommands = commandRegistry.getCommands().filter { !it.requiresChatAdmin() && !it.requiresAppCeo() }
        val adminCommands = commandRegistry.getCommands().filter { it.requiresChatAdmin() && !it.requiresAppCeo() }
        val ceoCommands = commandRegistry.getCommands().filter { it.requiresAppCeo() }
        if (userCommands.isNotEmpty()) {
            userCommands.forEach {
                builder.append("/sweet ${it.getUsage()}\n")
            }
        }
        if (adminCommands.isNotEmpty()) {
            builder.append("\nДля админов:\n")
            adminCommands.forEach {
                builder.append("/sweet ${it.getUsage()}\n")
            }
        }
        if (ceoCommands.isNotEmpty()) {
            builder.append("\nДля CEO:\n")
            ceoCommands.forEach {
                builder.append("/sweet ${it.getUsage()}\n")
            }
        }

        vkMessagesService.send(message.peerId, builder.toString())
    }
}